### DISCUSSION TITLE
Show emojis below elements

### DISCUSSION THREAD

**Discussion by DarkMagus** [clientIO/joint#2700] [Q&A]
### Introduction

I would like to show emojis (or tags as we called them here) placed just below the graph elements. The user can add/remove as many as they wish as shown in the below image:

![image](https://github.com/clientIO/joint/assets/4059723/fe3b54da-f535-4e40-adeb-1d020bab1de7)

When the user hover on specific emoji, a tooltip is shown with associated name as show below:

![image](https://github.com/clientIO/joint/assets/4059723/86ba9c9d-ac05-48c2-b035-e57652416e72)

What is the best way to achieve the above using JointJS+?  
I am thinking about elementTools but I am already using them elsewhere in my app, and from what I understood elementTools can be hidden and shown together, therefore will interfere with the already implemented elementTools

Thanks lot

### Does your question relate to JointJS or JointJS+. Select both if applicable.

JointJS+

**Comment by kumilingus** [comment]
Use the [`list` highlighter](https://docs.jointjs.com/api/highlighters/#list).

Examples:
https://www.jointjs.com/demos/icons
https://www.jointjs.com/demos/dynamic-status-icons

You can use [`ui.Stencil`](https://docs.jointjs.com/learn/features/element-palette) to drag and drop icons on the elements (intercept the default behavior and add the new icon instead).

Example:
https://www.jointjs.com/demos/drop-stencil-element-as-shape-icon

**Reply by DarkMagus** [reply]
@kumilingus 
How can I draw list highlighter at right-bottom but a bit outside the element like in my figures?
Thanks

**Reply by kumilingus** [reply]
Using `position` and `margin` options:
```ts
{
  position: "bottom-right",
  margin: { right: 0, bottom: -25 },
}
```

**Reply by DarkMagus** [reply]
@kumilingus 
I will apply your recommendations and let you know on my findings.
Thanks lot Roman.

**Reply by DarkMagus** [reply]
@kumilingus 
```element.prop('myattribute', 'myvalue')``` adds ```myattribute``` with value equal to ```myvalue``` to element attributes.

is there an equivalent method that does the opposite? I mean something that removes ```myattribute``` from the element attributes

Thanks

**Reply by kumilingus** [reply]
Yes, you can use  [element.removeProp()](https://docs.jointjs.com/api/dia/Cell/#removeprop).

**Reply by DarkMagus** [reply]
@kumilingus 
I used your recommendations and here is the version of the above figures using list highlighter:

![image](https://github.com/clientIO/joint/assets/4059723/a84b8c08-564c-4273-8ac1-acbe14399c8c)

The code to create and display the tags is written below. Let me know if there is a better way to improve it! Thanks Roman again for your recommendations and advises. I appreciate!
```
class TagsEffect extends joint.highlighters.list {
    
    createListItem(tag: Tag, { width, height }: joint.dia.Size, currentTag: SVGElement): SVGElement {
        let listItem = currentTag;
        if (!listItem) {
            listItem = document.createElementNS('http://www.w3.org/2000/svg', 'g');
        }
        listItem.setAttribute('cursor', 'pointer');
        
        listItem.setAttribute('id', 'tag-' + tag.id!.toString());
        listItem.setAttribute('class', 'tag');

        
        listItem.innerHTML = `
            <rect x=0 y=0 width="${width}" height="${height}" fill="none" stroke="black" stroke-width="1" />
            <text x="${width / 2}" y="${height / 2}" dominant-baseline="middle" text-anchor="middle" >${tag.presentation.unicode}</text>
        `;

        const tooltip = new joint.ui.Tooltip({
            root: listItem,
            target: listItem,
            content: tag.name,
            position: joint.ui.Tooltip.TooltipPosition.Top,
            padding: 10,
            theme: 'light',
            delay: 200,
        });
        
        tooltip.hide();
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        listItem.addEventListener('mouseenter', (evt) => {
            // eslint-disable-next-line no-console
            console.log('hover:', tag.name);
            tooltip.show();
        });

        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        listItem.addEventListener('mouseenter', (evt) => {
            // eslint-disable-next-line no-console
            console.log('hover:', tag.name);
            tooltip.hide();
        });

        return listItem;
    }
    
}

function adjustTagUnicode(tag: Tag): Tag {
    let codes = tag.presentation.unicode!.split('-').map((code: string) => parseInt(`0x${code})`));
    let adjustedUnicode = String.fromCodePoint(...codes);
    let newPresentation = { ...tag.presentation, unicode: adjustedUnicode };
    return { ...tag, presentation: newPresentation };
}

export function setTagIcons(paper: joint.dia.Paper, element: joint.dia.Element, tags: Tag[]): void {
    element.removeProp('tags');
    tags.forEach((tag, index) => {
        element.prop(['tags', Number(index)], adjustTagUnicode(tag));
    });

    TagsEffect.add(element.findView(paper), 'root', 'tags', {
        attribute: 'tags',
        position: 'bottom-right',
        margin: { right: 0, bottom: -40 },
        size: 20,
        gap: 5
        
    })
}




```

**Reply by kumilingus** [reply]
You shouldn't create a new instance of `ui.Tooltip` for each icon (you never remove the instances which would lead to memory leaks).

It's better to add `listItem.dataset.tooltip = tag.name;` and configure a single `ui.Tooltip` with `data-tooltip` target.

```ts
const tooltip = new ui.Tooltip({
    target: '[data-tooltip]',
    direction: 'top',
    padding: 10
});
```

**Reply by DarkMagus** [reply]
@kumilingus -  I applied your above recommendation but when the mouse is over an icon belonging to an element with several icons, it always shows the tooltip of the first icon. For example, the mouse is over the cold icon, but the tooltip of Winking eye icon is displayed instead

![image](https://github.com/clientIO/joint/assets/4059723/203aa379-0432-48fe-aed8-d10fdecf3260)

Here is the modified code with a single ui.Tooltip:

```
const tooltip = new joint.ui.Tooltip({
    target: '[data-tooltip]',
    position: joint.ui.Tooltip.TooltipPosition.Top,
    padding: 10,
});

class TagsEffect extends joint.highlighters.list {
    
    createListItem(tag: Tag, { width, height }: joint.dia.Size, currentTag: SVGElement): SVGElement {
        let listItem = currentTag;
        if (!listItem) {
            listItem = document.createElementNS('http://www.w3.org/2000/svg', 'g');
        }
        listItem.setAttribute('cursor', 'pointer');
        
        listItem.setAttribute('id', 'tag-' + tag.id!.toString());
        listItem.setAttribute('class', 'tag');
        
        listItem.innerHTML = `
            <rect x=0 y=0 width="${width}" height="${height}" fill="none" stroke="black" stroke-width="1" />
            <text x="${width / 2}" y="${height / 2}" dominant-baseline="middle" text-anchor="middle" >${tag.presentation.unicode}</text>
        `;


        listItem.dataset.tooltip = tag.name;

        tooltip.hide();
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        listItem.addEventListener('mouseenter', (evt) => {
            // eslint-disable-next-line no-console
            console.log('hover:', tag.name);
            tooltip.show();
        });

        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        listItem.addEventListener('mouseout', (evt) => {
            // eslint-disable-next-line no-console
            console.log('hover:', tag.name);
            tooltip.hide();
        });

        return listItem;
    }
    
}

```

**Reply by kumilingus** [reply]
This should do it:

```ts
const tooltip = new joint.ui.Tooltip({
    target: '[data-tooltip]',
    position: joint.ui.Tooltip.TooltipPosition.Top,
    padding: 10,
});

class TagsEffect extends joint.highlighters.list {
    
    createListItem(tag: Tag, { width, height }: joint.dia.Size, currentTag: SVGElement): SVGElement {
        let listItem = currentTag;
        if (!listItem) {
            listItem = document.createElementNS('http://www.w3.org/2000/svg', 'g');
        }
        listItem.setAttribute('cursor', 'pointer');
        
        listItem.setAttribute('id', 'tag-' + tag.id!.toString());
        listItem.setAttribute('class', 'tag');
        
        listItem.innerHTML = `
            <rect x=0 y=0 width="${width}" height="${height}" fill="none" stroke="black" stroke-width="1" />
            <text x="${width / 2}" y="${height / 2}" dominant-baseline="middle" text-anchor="middle" >${tag.presentation.unicode}</text>
        `;

        listItem.dataset.tooltip = tag.name;

        return listItem;
    }
    
}
```

**Reply by DarkMagus** [reply]
@kumilingus 
I tried your code but the tooltip does not show up at all. Is there a reason for removing the ```mouseenter``` and ```mouseout``` event listners?

Thanks Roman.

**Reply by kumilingus** [reply]
The `Tooltip` is already listening for `mouseenter` and `mouseleave` events.
Any element in the DOM that has the `data-tooltip` attribute will display a tooltip when hovered over.

**Reply by DarkMagus** [reply]
@kumilingus 
I cannot make it work with your code, it looks like Tooltip just not listening to mouse events

**Reply by kumilingus** [reply]
If you post here a complete code I can take a look.

**Reply by DarkMagus** [reply]
@kumilingus I appreciate your proposal; however, it will require a clearance from the sponsor which might take ages.
I added back to your code the two mouse events and it seems to work now with only ```one single ui Tooltip``` for all icons. Here is the modified code snippet. See the mouseenter event listener

```
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const tooltip = new joint.ui.Tooltip({
    target: '[data-tooltip]',
    position: joint.ui.Tooltip.TooltipPosition.Top,
    padding: 10,
});



class TagsEffect extends joint.highlighters.list {
    
    createListItem(tag: Tag, { width, height }: joint.dia.Size, currentTag: SVGElement): SVGElement {
        let listItem = currentTag;
        if (!listItem) {
            listItem = document.createElementNS('http://www.w3.org/2000/svg', 'g');
        }
        listItem.setAttribute('cursor', 'pointer');
        
        listItem.setAttribute('id', 'tag-' + tag.id!.toString());
        listItem.setAttribute('class', 'tag');
        
        listItem.innerHTML = `
            <rect x=0 y=0 width="${width}" height="${height}" fill="none" stroke="black" stroke-width="1" />
            <text x="${width / 2}" y="${height / 2}" dominant-baseline="middle" text-anchor="middle" >${tag.presentation.unicode}</text>
        `;

        listItem.dataset.tooltip = tag.name;

        listItem.addEventListener('mouseenter', (evt: MouseEvent) => {
            const target = evt.target as SVGElement;
            tooltip.options.target = target;
            tooltip.options.content = target.dataset.tooltip;
            
            tooltip.show();
        });

        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        listItem.addEventListener('mouseout', (evt) => {
            tooltip.hide();
        });        

        return listItem;
    }
    
}


```

**Reply by DarkMagus** [reply]
@kumilingus 
I did several tests and the above code with the addition of the event listeners and a single tooltip instance seems to work well.
Thanks again for your patience and advice

**Reply by kumilingus** [reply]
It probably works fine, but you add 2 event listeners for each icon as opposed to the constant 2 listeners created by the tooltip. It's not very memory efficient.

I didn't mean for you to share your source code. Provide us with an example code that demonstrates the problem, and anyone can simply copy it and run it.

**Reply by DarkMagus** [reply]
@kumilingus You are right regarding the memory consumption specially with regard to the use of anonymous handler functions. There is a way that results in smaller memory consumption by using a normal function instead of anonymous handler functions.
See Memory issues here https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener

The tooltip handling event is just not working for me. I will try to create a simple react project to outline the issue and send it to you.

Thanks

**Reply by DarkMagus** [reply]
@kumilingus 
Hi Roman -  I found why the tooltip did not work. By adding ```rootTarget``` as follow:

```
const tooltip = new joint.ui.Tooltip({
    rootTarget: document.body,
    target: '[data-tooltip]',
    position: joint.ui.Tooltip.TooltipPosition.Top,
    padding: 10,
});

```

with rootTarget being added, I no longer need the addEventListner for every listItem

**Reply by kumilingus** [reply]
Great! My bad for leaving that out of my suggestion.

**Reply by DarkMagus** [reply]
Thanks lot Roman for your advices and support. I do not regret advising my client to buy JointJS+ licenses

**Comment by DarkMagus** [comment]
@kumilingus 
Hi Roman - Is it possible to move an emoji around the element?

Thanks

**Reply by kumilingus** [reply]
What do you mean? Do you want the user to be able to drag the emoji over the element (then why?), or to be able to copy the emoji by dropping it on other elements?
Please elaborate.

**Reply by DarkMagus** [reply]
I want the user to be able to drag the emoji over the element. why? this is a user requirements to enable users to place the emojis wherever they want

**Reply by kumilingus** [reply]
Then you need to write a custom highlighter that renders multiple emojis and implements drag & drop.
Or you switch to represent emojis as elements and leverage the cell embedding (embed multiple emojis into parent element).
